mlflow/store/s3_artifact_repo.py

```diff
@@ -69,15 +69,29 @@ def list_artifacts(self, path=None):
         for result in results:
```

Comment:
Let's write some sort of a checker to make sure that nesting is legit. For instance it is possible to have this sort of directory structure in S3

```
dir_name                # is a true directory
dir_name/sub_dir        # this one is a file
dir_name/sub_dir/file   # this is also a file
```

Reply:
After offline discussion, it appears that we don't currently have a mechanism for checking / enforcing that S3 object keys containing slashes are directories. We should definitely agree on a strategy for dealing with files whose keys contain slashes, but this is a bit outside the scope of the current PR.